### PR TITLE
Miner actor related

### PR DESCRIPTION
**Pull request by turuslan**
### Description of the Change
Miner actor related:
- Market actor types
- Storage power actor types
- Primitive types
- Proof sector types
- Runtime methods
- ACTOR_METHOD macro
- Miner actor types
- Miner actor policy constants

### Benefits
Split miner actor PR into smaller parts.

### Possible Drawbacks 
None

### CODE REVIEW

File: core/proofs/sector.hpp [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+  using SealRandomness = Randomness;
+
+  using InteractiveRandomness = Randomness;
+
+  struct SealProof {
+    Buffer proof;
+  };
+
+  struct OnChainSealVerifyInfo {
```
Comment (Alexey-N-Chernyshov) [inline core/proofs/sector.hpp]: Structures are partly commented in specs-actors, please leave comment for structures where it is possible.

File: core/storage/hamt/hamt.cpp [hunk 2]
```diff
@@ -88,6 +88,10 @@ namespace fc::storage::hamt {
 
   outcome::result<CID> Hamt::flush() {
     OUTCOME_TRY(flush(root_));
+    return cid();
```
Comment (Alexey-N-Chernyshov) [inline core/storage/hamt/hamt.cpp]: Why? You will always use cid after flush.

File: core/storage/hamt/hamt.hpp [hunk 3]
```diff
@@ -160,6 +160,9 @@ namespace fc::storage::hamt {
      */
     outcome::result<CID> flush();
 
+    /// Get root CID if flushed, throw otherwise
+    const CID &cid() const;
```
Comment (Alexey-N-Chernyshov) [inline core/storage/hamt/hamt.hpp]: Why do you need it?

File: core/vm/actor/builtin/miner/types.hpp [hunk 4]
```diff
@@ -74,6 +83,67 @@ namespace fc::vm::actor::builtin::miner {
     PoStState post_state;
   };
 
+  enum class CronEventType {
```
Comment (Alexey-N-Chernyshov) [inline core/vm/actor/builtin/miner/types.hpp]: Please, leave comments where it is possible.

File: core/vm/runtime/runtime.hpp [hunk 5]
```diff
@@ -141,6 +142,59 @@ namespace fc::vm::runtime {
...
+      return codec::cbor::decode<R>(result.return_value);
+    }
+
+    /// Send with typed params P and result R
+    template <typename R, typename P>
+    outcome::result<R> sendPR(Address to_address,
+                              MethodNumber method_number,
+                              const P &params,
+                              BigInt value) {
+      OUTCOME_TRY(params2, codec::cbor::encode(params));
```
Comment (Alexey-N-Chernyshov) [inline core/vm/runtime/runtime.hpp]: encodeActorParams

File: core/vm/runtime/runtime.hpp [hunk 6]
```diff
@@ -141,6 +142,59 @@ namespace fc::vm::runtime {
...
+                                            MethodNumber method_number,
+                                            const P &params,
+                                            BigInt value) {
+      OUTCOME_TRY(params2, codec::cbor::encode(params));
+      return send(to_address, method_number, MethodParams{params2}, value);
+    }
+
+    /// Get decoded current actor state
+    template <typename T>
+    outcome::result<T> getCurrentActorStateCbor() {
```
Comment (Alexey-N-Chernyshov) [inline core/vm/runtime/runtime.hpp]:
```suggestion
    outcome::result<T> getCurrentActorState() {
```

File: core/vm/runtime/runtime.hpp [hunk 7]
```diff
@@ -141,6 +142,59 @@ namespace fc::vm::runtime {
...
+    }
+
+    /// Send with typed result R
+    template <typename R>
+    outcome::result<R> sendR(Address to_address,
+                             MethodNumber method_number,
+                             const MethodParams &params,
+                             BigInt value) {
+      OUTCOME_TRY(result, send(to_address, method_number, params, value));
+      return codec::cbor::decode<R>(result.return_value);
```
Comment (Alexey-N-Chernyshov) [inline core/vm/runtime/runtime.hpp]: decodeActorParams

File: core/vm/runtime/runtime.hpp [hunk 8]
```diff
@@ -141,6 +142,59 @@ namespace fc::vm::runtime {
...
+
+    /// Send funds
+    inline auto sendFunds(const Address &to, BigInt value) {
+      // kSendMethodNumber circular dependency
+      return send(to, {0}, {}, value);
+    }
+
+    /// Send with typed result R
+    template <typename R>
+    outcome::result<R> sendR(Address to_address,
```
Comment (Alexey-N-Chernyshov) [inline core/vm/runtime/runtime.hpp]: I don't like the sendPR methods. We have encodeActorParams/decodeActorParams functions, and these methods seems to be an overkill. Names are a bit tricky (sendPR). For me these methods do not make code more readable and simpler.

File: core/vm/exit_code/impl/exit_code.cpp [hunk 9]
```diff
@@ -40,6 +38,20 @@ namespace fc::vm {
...
+      case E::MINER_ACTOR_ILLEGAL_ARGUMENT:
+        return 16;  // ErrIllegalArgument in actor-specs
+      case E::MINER_ACTOR_WRONG_CALLER:
+        return 18;  // ErrForbidden in actor-specs
+      case E::MINER_ACTOR_POST_TOO_LATE:
+      case E::MINER_ACTOR_POST_TOO_EARLY:
+      case E::MINER_ACTOR_ILLEGAL_STATE:
+        return 20;  // ErrIllegalState in actor-specs
+      case E::MINER_ACTOR_INSUFFICIENT_FUNDS:
+        return 19;  // ErrInsufficientFunds in actor-specs
```
Comment (Alexey-N-Chernyshov) [inline core/vm/exit_code/impl/exit_code.cpp]: please, reorder in ascendance order.

File: core/storage/amt/amt.hpp [hunk 10]
```diff
@@ -140,8 +140,12 @@ namespace fc::storage::amt {
     outcome::result<Value> get(uint64_t key);
     /// Remove value by key, does not write to storage
     outcome::result<void> remove(uint64_t key);
+    /// Checks if key is present
+    outcome::result<bool> contains(uint64_t key);
     /// Write changes made by set and remove to storage
     outcome::result<CID> flush();
+    /// Get root CID if flushed, throw otherwise
+    const CID &cid() const;
```
Comment (Alexey-N-Chernyshov) [inline core/storage/amt/amt.hpp]: The same as for HAMT. Why, when we use it with flush().

File: core/vm/interpreter/interpreter.cpp [hunk 11]
```diff
@@ -182,11 +181,11 @@ namespace fc::vm::interpreter {
     for (auto i = 0u; i < receipts.size(); ++i) {
       OUTCOME_TRY(receipts_amt.setCbor(i, receipts[i]));
     }
-    OUTCOME_TRY(receipts_root, receipts_amt.flush());
+    OUTCOME_TRY(receipts_amt.flush());
 
     return Result{
         new_state_root,
-        receipts_root,
+        receipts_amt.cid(),
```
Comment (Alexey-N-Chernyshov) [inline core/vm/interpreter/interpreter.cpp]: ?

File: core/vm/runtime/impl/runtime_impl.cpp [hunk 12]
```diff
@@ -180,6 +171,23 @@ fc::outcome::result<void> RuntimeImpl::transfer(Actor &from,
...
 }
 
+fc::outcome::result<Address> RuntimeImpl::resolveAddress(
+    const Address &address) {
+  return state_tree_->lookupId(address);
+}
+
+fc::outcome::result<bool> RuntimeImpl::verifyPoSt(uint64_t sector_size,
+                                                  const PoStVerifyInfo &info) {
+  // TODO(turuslan): FIL-112 connect verifyPoSt from proofs
```
Comment (artyom-yurin) [inline core/vm/runtime/impl/runtime_impl.cpp]: Please change to FIL-160

File: core/vm/runtime/impl/runtime_impl.cpp [hunk 13]
```diff
@@ -180,6 +171,23 @@ fc::outcome::result<void> RuntimeImpl::transfer(Actor &from,
...
+  return RuntimeError::UNKNOWN;
+}
+
+fc::outcome::result<bool> RuntimeImpl::verifySeal(uint64_t sector_size,
+                                                  const SealVerifyInfo &info) {
+  // TODO(turuslan): FIL-112 connect verifySeal from proofs
```
Comment (artyom-yurin) [inline core/vm/runtime/impl/runtime_impl.cpp]: Please change to FIL-160

File: core/primitives/chain_epoch/chain_epoch_codec.cpp [hunk 14]
```diff
@@ -11,11 +11,15 @@ namespace fc::primitives::chain_epoch {
...
+    auto encoded = uvarint.toBytes();
+    return std::string(encoded.begin(), encoded.end());
+  }
+
   std::string encodeToByteString(const ChainEpoch &epoch) {
     // TODO (a.chernyshov) actor-specs uses Protobuf Varint encoding
-    UVarint number(epoch);
-    auto encoded = number.toBytes();
-    return std::string(encoded.begin(), encoded.end());
+    return uvarintKey(epoch);
```
[the remark this answers is not in the document]
Comment (turuslan) [reply core/primitives/chain_epoch/chain_epoch_codec.cpp]: They are not. You commented that epoch encodeToByteString is `varintKey`, not `uvarintKey`

File: core/storage/hamt/hamt.cpp (hunk 2 above)
Comment (turuslan) [reply core/storage/hamt/hamt.cpp]: Left for compatibility.
New code will not create new names for flushed cid and just refer amt/hamt object.

File: core/storage/hamt/hamt.hpp (hunk 3 above)
Comment (turuslan) [reply core/storage/hamt/hamt.hpp]: To get root cid separately from flush, as OUTCOME_TRY requires new name for flushed cid

File: core/vm/runtime/runtime.hpp (hunk 6 above)
Comment (turuslan) [reply core/vm/runtime/runtime.hpp]: Still need way to get cid, when type is unknown

File: core/vm/runtime/runtime.hpp (hunk 7 above)
Comment (turuslan) [reply core/vm/runtime/runtime.hpp]: Wrong, this is result, not params

File: core/vm/runtime/runtime.hpp (hunk 8 above)
Comment (turuslan) [reply core/vm/runtime/runtime.hpp]: encodeActorParams and decoding result would use two extra OUTCOME_TRY, with need for 2 extra variable names.
Hiding encoding and decoding into templates seems better to me

File: core/crypto/randomness/randomness_types.hpp [hunk 15]
```diff
@@ -21,7 +21,9 @@ namespace fc::crypto::randomness {
   enum class DomainSeparationTag : size_t {
     TicketDrawingDST = 1,
     TicketProductionDST = 2,
-    PoStDST = 3
+    PoStDST = 3,
+    SealRandomness,
```
Comment (igor-egorov) [inline core/crypto/randomness/randomness_types.hpp]: Please define values for the newly added items for consistency purposes

File: core/vm/actor/actor_method.hpp [hunk 16]
```diff
@@ -12,6 +12,10 @@
 #include "vm/exit_code/exit_code.hpp"
 #include "vm/runtime/runtime.hpp"
 
+#define ACTOR_METHOD(name)                \
```
Comment (igor-egorov) [inline core/vm/actor/actor_method.hpp]: cool!
🚀

**Pull request merged**